### ISSUE TITLE
Assess integration with KubeVirt

### ISSUE THREAD

**Issue by sboeuf** [cloud-hypervisor/cloud-hypervisor#4201]
After some research, here is a quick plan on how Cloud Hypervisor could be integrated with KubeVirt. This will be a prototype to prove it's feasible.

The right abstraction layer seems to be the `virt-launcher`, as it is spawned for each VM to be created. There are some good things from the __libvirt__ implementation that can be reused in order to avoid rewriting everything from scratch. In particular, using the interface `DomainManager` as the abstraction layer to implement most of the interaction with Cloud Hypervisor seems the right thing to do. The `virt-launcher` binary will have to be Cloud Hypervisor specific as well since it would be too hard to find a proper factorization for everything between __libvirt__ and Cloud Hypervisor.

#### Steps to add Cloud Hypervisor basic support to KubeVirt
- Add autogenerated openapi go package about Cloud Hypervisor
- Start Cloud Hypervisor with the API socket being a random socket path, and use a socketpair for `event-monitor`
- Provide the API socket path to the `DomainManager` creation so that it can initialize the CH openapi client
- Implement the `DomainManager` interface (start with `SyncVMI`, `PauseVMI`, ...), but leave migration related and guest related methods empty for now
- Implement a package for converting `v1.VirtualMachineInstance` structure into Cloud Hypervisor configuration (this will be mostly used from `SyncVMI`)
- Create a thread for handling events from the `event-monitor` and send them up to the `virt-handler` (`HandleDomainEvent` method from the `Notify` interface)

The goal here is to be able to create and run a VM from `kubectl`, and connect to the console with `virtctl` to access the guest.

#### Further improvements
- Look into networking to be able to create a VM with one or more network interfaces, which can be helpful for outside connectivity as well as for SSHing into the VM
- Implement an agent for Cloud Hypervisor (#3196) in order to fulfill the expectations from KubeVirt and be able to implement things like `GuestOSInfo`, `Exec`, `GuestPing`, `InterfacesStatus`, `GetFilesystems` and `GetUsers`
- Relying on the new agent, implement the missing methods in the `DomainManager` implementation
- Once everything is working as expected, we can start tackling the migration support, by implementing the migration related methods in `DomainManager`

**Comment by yuchen-sun** [comment]
👍 
We are always looking for opportunities to add CH support to our Kubevirt cluster. We think CH is more presentational and better suited to our application.

**Comment by sboeuf** [comment]
> +1 We are always looking for opportunities to add CH support to our Kubevirt cluster. We think CH is more presentational and better suited to our application.

Glad to hear that!

**Comment by sdake** [comment]
@sboeuf qubevirt is one solution, however, I feel that we need a broader solution for managing a node with a boatload of `chv` running. The fundamental technical problem with qubevirt is that when Kubernetes runs on bare metal, and the bare metal has, for example, 2 terabytes of memory, Kubernetes has a published limit of 8 bits worth of pods (110 pods). And Kubernetes works much better with 5 bits worth of pods (30 pods). As you can see, increasing the bit count to represent pods causes Kubernetes to, for lack of a better word, completely and totally implode. Running 300 pods with overrides to the defaults of Kubernetes results in a nonusable K8s, even on a bare metal machine such as a Cisco UCS C220 M5 with 2 Gold Xeon CPUs and 2tb of memory.

And even with 300 pods, I find it difficult to use much more than a few hundred gigs of ram. Modern servers ship with 2TB...

I am not indicating qubevirt is bad or K8s is bad, just that this is not the only technical deployment model, and probably a poor one.

Why not consider developing, at least a virtualization API, to manage a fleet of chv with random socket IDs with no configuration, that can be configured on the fly? (In fact I am working on such an API now implemented in rust, hence the question). And I would love to join forces with others in this endeavor.
Cheers,
-steve

**Comment by sboeuf** [comment]
> @sboeuf qubevirt is one solution, however, I feel that we need a broader solution for managing a node with a boatload of `chv` running. The fundamental technical problem with qubevirt is that when Kubernetes runs on bare metal, and the bare metal has, for example, 2 terabytes of memory, Kubernetes has a published limit of `7` bits worth of pods (110 pods). And Kubernetes works much better with `5` bits worth of pods (30 pods). As you can see, increasing the bit count to represent pods causes Kubernetes to, for lack of a better word, completely and totally implode. Running 300 pods with overrides to the defaults of Kubernetes results in a nonusable K8s, even on a bare metal machine such as a Cisco UCS C220 M5 with 2 Gold Xeon CPUs and 2tb of memory.
> 
> And even with 300 pods, I find it difficult to use much more than a few hundred gigs of ram. Modern servers ship with 2TB...
> 
> I am not indicating qubevirt is bad or K8s is bad, just that this is not the only technical deployment model, and probably a poor one.

Interesting feedback, very valuable. Have you been able to identify what's making k8s "implode" in such situation? Kubernetes is a big open source project and when used out of the box I don't think it meets production standards. A lot of companies rely on their engineers to overcome that kind of limitations, and they don't necessarily take the time to contribute back to the upstream k8s project.

> 
> Why not consider developing, at least a virtualization API, to manage a fleet of chv with random socket IDs with no configuration, that can be configured on the fly? (In fact I am working on such an API now implemented in rust, hence the question). And I would love to join forces with others in this endeavor. Cheers, -steve

This sounds like a very nice initiative but this is something that can be done in parallel to the KubeVirt integration effort. The hard bit with your proposal is about getting the API adopted by different communities so that it can be integrated in other projects. This takes time and we never know if it will end up being adopted. I can see this as a long term effort.

The KubeVirt integration is more something like a mid term effort and the benefit can be quite immediate for k8s users as they would be able to run both secure containers and VMs relying on Cloud Hypervisor (using Kata Containers and KubeVirt respectively).

**Comment by fengye87** [comment]
Hi, we've actually made a project named [Virtink](https://github.com/smartxworks/virtink) to support Cloud Hypervisor VMs solely on Kubernetes with some cool features like direct kernel boot with container rootfs.

We like KubeVirt and think it's a great work to manage the gap between (traditional) VM and containerized workload. However, we feel like it's both unnatural and unwieldy to port KubeVirt with Cloud Hypervisor. We actually tried this road in one of our projects but then decided it's not a good idea for following reasons:

1. VMI validation needs to be branched a lot to deal with CH limitations, like UEFI only, limited device bus, no VNC...
2. A whole different launcher needs to be created, probably without libvirt
3. Kubernetes controllers should do one thing and do it well

So what's why we created Virtink from scratch to support Cloud Hypervisor VMs solely on Kubernetes. It takes off the burden of libvirtd and the long-running launcher process so that it's more lightweight. And building on CH's direct kernel boot ability, Virtink could take a container image as a VM rootfs and direct boot into it. Our sight of it is that Virtink and KubeVirt can be two different but complement Kubernetes addons, KubeVirt is more specialized in traditional virtualized workloads while Virtink can be more specialized in cloud workloads.

Checkout our [Virtink](https://github.com/smartxworks/virtink) project if you are interested. Feel free to give us any feedbacks or questions on the project's issue board.

**Comment by sboeuf** [comment]
> Hi, we've actually made a project named [Virtink](https://github.com/smartxworks/virtink) to support Cloud Hypervisor VMs solely on Kubernetes with some cool features like direct kernel boot with container rootfs.

Nice to hear about that!

> 
> We like KubeVirt and think it's a great work to manage the gap between (traditional) VM and containerized workload. However, we feel like it's both unnatural and unwieldy to port KubeVirt with Cloud Hypervisor. We actually tried this road in one of our projects but then decided it's not a good idea for following reasons:
> 
>     1. VMI validation needs to be branched a lot to deal with CH limitations, like UEFI only, limited device bus, no VNC...

True, but that's also the point of this effort, showing where KubeVirt makes too many assumptions about things being supported assuming it's using libvirt and make it more flexible.

> 
>     2. A whole different launcher needs to be created, probably without libvirt

Yes that's exactly the approach I took with the POC posted here: https://github.com/kubevirt/kubevirt/pull/8056

> 
>     3. Kubernetes controllers should do one thing and do it well
> 
> 
> So what's why we created Virtink from scratch to support Cloud Hypervisor VMs solely on Kubernetes. It takes off the burden of libvirtd and the long-running launcher process so that it's more lightweight. And building on CH's direct kernel boot ability, Virtink could take a container image as a VM rootfs and direct boot into it. 

>Our sight of it is that Virtink and KubeVirt can be two different but complement Kubernetes addons, KubeVirt is more specialized in traditional virtualized workloads while Virtink can be more specialized in cloud workloads.

I fully agree with that statement, I feel like Virtink and KubeVirt can serve two different purposes, and I'm glad they can both provide a way for our users to easily manage Cloud Hypervisor's VMs.

> 
> Checkout our [Virtink](https://github.com/smartxworks/virtink) project if you are interested. Feel free to give us any feedbacks or questions on the project's issue board.

:+1:

**Comment by fengye87** [comment]
Thank you @sboeuf , we are motivated. Virtink is still at its early stage of development. Hope we can get more feedbacks from CH community to keep the project improving.

**Comment by rbradford** [comment]
What are the next steps on this issue? Is the assessment complete?

**Comment by sboeuf** [comment]
Yes the assessment is complete. The KubeVirt community doesn't want to pursue with this approach as they would like to see a new abstraction layer being created for supporting multiple VMMs. We won't do that.

**Comment by emirb** [comment]
_(The early bird gets the worm)_

Now that https://www.cncf.io/blog/2026/03/25/announcing-the-release-of-kubevirt-v1-8/ shipped a Hypervisor Abstraction Layer (announced two days ago), is it worth revisiting this?                                                                            
                                                        
In my understanding the community wanted a proper abstraction layer before supporting additional VMMs, and understandably nobody wanted to build that from the CH side. That abstraction layer now exists upstream.

**Comment by sboeuf** [comment]
> _(The early bird gets the worm)_
> 
> Now that https://www.cncf.io/blog/2026/03/25/announcing-the-release-of-kubevirt-v1-8/ shipped a Hypervisor Abstraction Layer (announced two days ago), is it worth revisiting this?
> 
> In my understanding the community wanted a proper abstraction layer before supporting additional VMMs, and understandably nobody wanted to build that from the CH side. That abstraction layer now exists upstream.

This would be quite nice! This would help grow CH usage and visibility from the k8s ecosystem.

**Comment by liuw** [comment]
> _(The early bird gets the worm)_
> 
> Now that https://www.cncf.io/blog/2026/03/25/announcing-the-release-of-kubevirt-v1-8/ shipped a Hypervisor Abstraction Layer (announced two days ago), is it worth revisiting this?
> 

That hypervisor abstraction layer is the work from my colleague for adding MSHV support. It is not really related to the VMM layer. The VMM is still QEMU.
